### PR TITLE
WIP - 58 implement transform adaptor

### PR DESCRIPTION
**Pull request by atbenmurray**
Fixes # .

### Description
The adaptor function that can be used to adapt non-comformant transforms or mixed combinations of tranforms when using Compose (see #59)

### Status
**Work in progress**

### Types of changes

- [x] Non breaking change (fix or new feature that would cause existing functionality to change)
- [x] New tests added to cover the changes
- [ ] Docstrings/Documentation updated

### CODE REVIEW

File: monai/data/transforms/adaptors.py [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+                        variable_name, types, type(variable)))
+
+    def must_be_types(variable_name, variable, types):
+        if not isinstance(variable, types):
+            raise ValueError(
+                "'{}' must be one of {} but is {}".format(
+                    variable_name, types, type(variable)))
+
+    def map_names(ditems, input_map):
+        return {k if k not in input_map else input_map[k]: v for k, v in ditems.items()}
```
Comment (ericspod) [inline monai/data/transforms/adaptors.py]: `k if k not in input_map else input_map[k]` -> `input_map.get(k,k)`

File: monai/data/transforms/adaptors.py [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+```
+class MyTransform1:
+    ...
+    def __call__(image):
+        return '''do stuff to image'''
+
+class MyTransform2:
+    ...
+    def __call__(img):
+        return '''do stuff to image'''
```
Comment (wyli) [inline monai/data/transforms/adaptors.py]: a typo: `to img` instead of `to image`

File: monai/data/transforms/adaptors.py [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+        import inspect
+        sfn = inspect.signature(function)
+        self.found_args = False
+        self.found_kwargs = False
+        self.defaults = {}
+        self.non_var_parameters = set()
+        for p in sfn.parameters.values():
+            if p.kind is inspect.Parameter.VAR_POSITIONAL:
+                self.found_args = True
+            if p.kind is inspect.Parameter.VAR_KEYWORD:
```
Comment (wyli) [inline monai/data/transforms/adaptors.py]: line 195 should be `elif` instead of `if`?

Comment (atbenmurray) [reply monai/data/transforms/adaptors.py]: Nope, it is correct. This code is setting flags if it found that parameter type present

File: monai/data/transforms/adaptors.py [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+                dinputs = map_only_names(ditems, inputs)
+
+        ret = function(**dinputs)
+
+        # now the mapping back to the output dictionary depends on outputs and what was returned from the function
+        op = outputs
+        if isinstance(ret, dict):
+            must_be_types_or_none('outputs', op, (dict,))
+            if op is not None:
+                ret = {v: ret[k] for k, v in op}
```
[the remark this answers is not in the document]
Comment (atbenmurray) [reply monai/data/transforms/adaptors.py]: It should be. I'll write tests for this code path; I thought I already had

File: monai/data/transforms/adaptors.py (hunk 2 above)
Comment (atbenmurray) [reply monai/data/transforms/adaptors.py]: Yes

File: monai/data/transforms/adaptors.py (hunk 1 above)
Comment (atbenmurray) [reply monai/data/transforms/adaptors.py]: I thought about it but came to the conclusion that input_map.get(k, k) would be less clear in its intent. I'm open to reevaluating.

Comment (ericspod) [reply monai/data/transforms/adaptors.py]: As we had discussed I think it's reasonable to expect people to know what `dict.get` does and that `get(k,k)` is understandable. The idiom `{remap.get(k,k):v for k,v in orig}` is understood to be a remapping of the keys in `orig`, where `remap` is a partial function from old keys to new. I'd say `if-else` expressions is more verbose and harder to read, and may be less obvious that it binds tighter than `:`.

File: monai/data/transforms/adaptors.py (hunk 3 above)
Comment (wyli) [reply monai/data/transforms/adaptors.py]: if the function input is 
```
def foo(image, *myargs): 
        pass
``` 
`self.non_var_parameters` will have `myargs` perhaps that's not the intended?

**Comment by atbenmurray** [comment]
Closing this pull request as it has been superceded by '58-implement-transform-adaptor-2'

**Pull request closed without merging**